src/packet/public_key_encrypted_session_key.rs

```diff
@@ -113,62 +247,172 @@ fn parse(
+                    key_version,
+                    fingerprint,
+                    pk_algo,
+                    values,
+                },
+            ))
+        } else {
+            Ok((
+                i,
+                PublicKeyEncryptedSessionKey::Other {
+                    packet_version,
+                    version,
+                },
+            ))
+        }
     }
 }
 
 impl Serialize for PublicKeyEncryptedSessionKey {
     fn to_writer<W: io::Write>(&self, writer: &mut W) -> Result<()> {
-        writer.write_u8(self.version)?;
-        writer.write_all(self.id.as_ref())?;
-        writer.write_u8(self.algorithm.into())?;
-
-        match self.algorithm {
-            PublicKeyAlgorithm::RSA
-            | PublicKeyAlgorithm::RSASign
-            | PublicKeyAlgorithm::RSAEncrypt
-            | PublicKeyAlgorithm::Elgamal
-            | PublicKeyAlgorithm::ElgamalSign => {
-                for mpi in &self.mpis {
-                    mpi.to_writer(writer)?;
+        writer.write_u8(self.version())?;
+
+        match self {
+            PublicKeyEncryptedSessionKey::V3 { id, .. } => writer.write_all(id.as_ref())?,
+            PublicKeyEncryptedSessionKey::V6 {
+                key_version,
+                fingerprint,
+                ..
+            } => {
+                // A one-octet size of the following two fields. This size may be zero, if the key version number field and the fingerprint field are omitted for an "anonymous recipient" (see Section 5.1.8).
+                match (key_version, fingerprint) {
+                    (Some(key_version), Some(fingerprint)) => {
+                        let len = fingerprint.len() + 1;
+                        writer.write_u8(len.try_into()?)?;
+
+                        // A one octet key version number.
+                        writer.write_u8((*key_version).into())?;
+
+                        // The fingerprint of the public key or subkey to which the session key is encrypted. Note that the length N of the fingerprint for a version 4 key is 20 octets; for a version 6 key N is 32.
+                        writer.write_all(fingerprint.as_bytes())?;
+                    }
+                    _ => writer.write_u8(0)?,
                 }
             }
-            PublicKeyAlgorithm::ECDH => {
-                self.mpis[0].to_writer(writer)?;
+            PublicKeyEncryptedSessionKey::Other { .. } => todo!(),
+        }
+
+        writer.write_u8(self.algorithm().into())?;
+
+        match (self.algorithm(), self.values()) {
+            (
+                PublicKeyAlgorithm::RSA
+                | PublicKeyAlgorithm::RSASign
+                | PublicKeyAlgorithm::RSAEncrypt
+                | PublicKeyAlgorithm::Elgamal
+                | PublicKeyAlgorithm::ElgamalSign,
+                _, // FIXME!
+            ) => {
+                // FIXME: impl Elgamal
+                let EskBytes::Rsa { mpi } = self.values() else {
+                    unimplemented!("FIXME")
```

Comment:
sad

Reply:
> sad

true. fixed in b737a345c42e746b4a73836f7a157d8bf373486c